src/liballoc/vec.rs

```diff
@@ -809,9 +809,13 @@ impl<T> Vec<T> {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn swap_remove(&mut self, index: usize) -> T {
-        let length = self.len();
-        self.swap(index, length - 1);
-        self.pop().unwrap()
+        unsafe {
+            // We replace self[index] with the last element. Note that this is
+            // safe even when index == self.len() - 1, as pop() only uses
+            // ptr::read and leaves the memory at self[index] untouched.
+            let hole: *mut T = &mut self[index];
+            ptr::replace(hole, self.pop().unwrap())
```

Comment:
I can't help but feel that it would be clearer and require less explanation to first move the element with `ptr::replace`, *then* decrease the size, rather than relying on the implementation of `pop`.

Reply:
I agree - in fact I had that implementation at some point. However in the discussion of https://github.com/rust-lang/rust/issues/52150 the emphasis was on less unsafe code, so I may have gone overboard a bit. I'll change it.

Reply:
I'm less concerned about the quantity, and more concerned that the unsafe code we *do* have should be as straightforward and obvious as possible.